source/common/buffer/buffer_impl.cc

```diff
@@ -94,7 +95,40 @@ void OwnedImpl::move(Instance& rhs, uint64_t length) {
 }
 
 int OwnedImpl::read(int fd, uint64_t max_length) {
-  return evbuffer_read(buffer_.get(), fd, max_length);
+  if (max_length == 0) {
+    return 0;
+  }
+  constexpr uint64_t MaxSlices = 2;
+  RawSlice slices[MaxSlices];
+  const uint64_t num_slices = reserve(max_length, slices, MaxSlices);
+  struct iovec iov[num_slices];
+  uint64_t num_slices_to_read = 0;
+  for (; num_slices_to_read < num_slices && max_length != 0; num_slices_to_read++) {
+    iov[num_slices_to_read].iov_base = slices[num_slices_to_read].mem_;
+    const size_t slice_length =
+        std::min(slices[num_slices_to_read].len_, static_cast<size_t>(max_length));
+    iov[num_slices_to_read].iov_len = slice_length;
+    ASSERT(max_length >= slice_length);
+    max_length -= slice_length;
+  }
+  auto& os_syscalls = Api::OsSysCallsSingleton::get();
+  const ssize_t rc = os_syscalls.readv(fd, iov, static_cast<int>(num_slices_to_read));
+  if (rc < 0) {
+    return rc;
```

Comment:
Can you quickly double check what evbuffer does when you reserve but don't commit? I think it's probably fine and the next reservation will use the allocated memory, but I'm not sure.

Reply:
Confirmed - in both the [single-buffer](https://github.com/libevent/libevent/blob/master/buffer.c#L1954) and [multi-buffer cases](https://github.com/libevent/libevent/blob/master/buffer.c#L2079), the next evbuffer reservation call uses the allocated memory.